### PR TITLE
Adding GPU support to Galaxy

### PR DESCRIPTION
**Pull request by gulsumgudukbay**
We are adding automated GPU support in Galaxy to run GPU-enabled tools. This effort is a part of the Galaxy academic project at Penn State funded by NSF. We are open to feedback and kindly want this to be pushed to the main upstream repository.

The changes can be tested manually (since there is no automated test for GPU job scheduling yet). We use the racon-gpu tool to test if the job is being scheduled to GPU/CPU correctly using test data provided within the package. If there is no GPU available at the host, the job will use the CPU-only executable of the racon tool automatically. Otherwise, CUDA-enabled executable will be used and the job will be run using GPU too. 

Note: this current commit is for the local runner with a single GPU host only. When this PR proceeds, I will also add the multi-GPU scheduling support.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Review by gulsumgudukbay — COMMENTED** [review]
fixed some basic cosmetic errors

**Review by gulsumgudukbay — COMMENTED** [review]
Fixed formatting for quotation marks

File: lib/galaxy/jobs/rules/destinations.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+out, err = sp.communicate()
+command_not_found = "command not found"
+if command_not_found.encode() not in out and command_not_found.encode() not in err:
+    # log.debug("Out: %s, err: %s, command_not_found.encode: %s COMMAND NOT FOUND NOT IN OUT" % (out, err, command_not_found.encode()))
+    gpu_flag = 1
+    import pynvml as nvml
+
+if gpu_flag == 1:
+    nvml.nvmlInit()
+    gpu_count = nvml.nvmlDeviceGetCount()
```
Comment (nuwang) [inline lib/galaxy/jobs/rules/destinations.py]: This segment of code is repeated in `local.py` and also `evaluation.py`. It can be refactored into a function. Also, there is a general design issue here - the decision to enable to disable a GPU should be made elsewhere. Clarification below.

File: lib/galaxy/jobs/rules/destinations.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def dynamic_fun(tool, job):
+    flag = 0
```
Comment (nuwang) [inline lib/galaxy/jobs/rules/destinations.py]: A dynamic destination is not the right place to determine whether GPUs should be enabled/disabled because dynamic destinations are optional.

You intent appears to be to process a new type of tool requirement for `gpus`. For a first pass, you could consider just moving the logic to `local.py`. Currently, administrators can specify a parameter named `local_slots` at the destination configuration here: https://github.com/galaxyproject/galaxy/blob/e8cb3d789c18ac5f8dc859fd58ebec31c4ef43f2/lib/galaxy/config/sample/job_conf.xml.sample_advanced#L445

That destination parameter is then processed here in `local.py`: ​https://github.com/galaxyproject/galaxy/blob/9143dd7ca46d150ebfb26febbe187979f682da51/lib/galaxy/jobs/runners/local.py#L58

You could add a similar destination parameter for gpus named `cuda_visible_devices`, and allow admins to explicitly list the CUDA_VISIBLE_DEVICES. Then in `local.py`, you can just use set the `CUDA_VISIBLE_DEVICES` env var based on that destination parameter. Note that destination parameters can also be overridden per tool in Galaxy, so this would allow admins fine control over GPU device visibility on a per tool basis.

Alternatively, admins can be allowed to just specify the the maximum number of gpus allowed at the destination through a parameter named `local_gpus`. Then, in `local.py`, you can perform `nvml.nvmlDeviceGetCount()`, and use the lower of the two values. Let's say that value is two. The code could then enumerate all available device_ids, and select the first two devices as `CUDA_VISIBLE_DEVICES`. This is of course, in the event that `cuda_visible_devices` destination parameter is not defined.

For a first pass, I would suggest avoiding changing the tool requirements file. Tools usually run fine even without a GPU, just slower, so allowing an administrator to configure this as a destination parameter seems more appropriate for starters.

Please also take a close look at the kubernetes runner that has similar parameters for controlling CPU and memory requests on a per tool basis: https://github.com/galaxyproject/galaxy/blob/e8cb3d789c18ac5f8dc859fd58ebec31c4ef43f2/lib/galaxy/config/sample/job_conf.xml.sample_advanced#L1061
What happens for GPUs should follow a similar pattern.

File: lib/galaxy/jobs/runners/local.py [hunk 3]
```diff
@@ -24,6 +24,20 @@
...
+out, err = sp.communicate()
+command_not_found = "command not found"
+if command_not_found.encode() not in out and command_not_found.encode() not in err:
+    # log.debug("LOCAL: out: %s, err: %s, command_not_found.encode: %s COMMAND NOT FOUND NOT IN OUT" % (out, err, command_not_found.encode()))
+    gpu_flag = 1
+    import pynvml as nvml
+
+if gpu_flag == 1:
+    nvml.nvmlInit()
+    gpu_count = nvml.nvmlDeviceGetCount()
```
Comment (nuwang) [inline lib/galaxy/jobs/runners/local.py]:
```suggestion
def get_local_gpu_count():
  bash_command = "/bin/bash -c 'nvidia-smi'"
  sp = subprocess.Popen(bash_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
  out, err = sp.communicate()
  command_not_found = "command not found"
  if command_not_found.encode() not in out and command_not_found.encode() not in err:
      # log.debug("LOCAL: out: %s, err: %s, command_not_found.encode: %s COMMAND NOT FOUND NOT IN OUT" % (out, err, command_not_found.encode()))
      import pynvml as nvml
      nvml.nvmlInit()
      return nvml.nvmlDeviceGetCount()
```

**Review by nuwang — CHANGES_REQUESTED** [review]
Thanks for continuing to improve on this.  The current implementation is a start, but it doesn't quite fit in correctly with how Galaxy handles resource requirements.  I've made some suggestions for how the code can be modified to make it more consistent with how Galaxy generally handles this.  @mvdbeek may have better suggestions.

File: lib/galaxy/jobs/runners/local.py (hunk 3 above)
Comment (gulsumgudukbay) [reply lib/galaxy/jobs/runners/local.py]: Hi, I addressed these comments in my latest commit. I also got rid of the usage of pynvml and did it with regular Python (so it won't need additional libraries). 
After this commit, I will work on adding the multi-gpu scheduling according to GPU memory usage.

**Comment by gulsumgudukbay** [comment]
Hi, I addressed these comments in my latest commit. I also got rid of the usage of pynvml and did it with regular Python (so it won't need additional libraries).
After this commit, I will work on adding the multi-gpu scheduling according to GPU memory usage.

File: lib/galaxy/jobs/runners/local.py [hunk 4]
```diff
@@ -49,23 +49,53 @@ def __init__(self, app, nworkers):
 
         super().__init__(app, nworkers)
 
+    def get_local_gpu_count(self):
+        bash_command = "/bin/bash -c 'nvidia-smi --query-gpu=index --format=csv'"
+        sp = subprocess.Popen(bash_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        out, err = sp.communicate()
+        command_not_found = "command not found"
+        if command_not_found.encode() not in out and command_not_found.encode() not in err:
+            data = (out.decode('utf-8').splitlines())[1:] #eliminate first element, as it is the header
```
Comment (nuwang) [inline lib/galaxy/jobs/runners/local.py]: Please use the python csv module (https://docs.python.org/3/library/csv.html) so that this value can be parsed more robustly.

File: lib/galaxy/jobs/runners/local.py [hunk 5]
```diff
@@ -49,23 +49,53 @@ def __init__(self, app, nworkers):
...
         # is going to work with other job runners.
         slots = job_wrapper.job_destination.params.get("local_slots") or os.environ.get("GALAXY_SLOTS")
+        cuda_visible_devices = job_wrapper.job_destination.params.get("cuda_visible_devices") or os.environ.get("CUDA_VISIBLE_DEVICES")
+        local_gpus = job_wrapper.job_destination.params.get("local_gpus") or os.environ.get("LOCAL_GPUS")
         if slots:
             slots_statement = f'GALAXY_SLOTS="{int(slots)}"; export GALAXY_SLOTS; GALAXY_SLOTS_CONFIGURED="1"; export GALAXY_SLOTS_CONFIGURED;'
         else:
             slots_statement = 'GALAXY_SLOTS="1"; export GALAXY_SLOTS;'
 
+        num_local_gpus = self.get_local_gpu_count()
```
Comment (nuwang) [inline lib/galaxy/jobs/runners/local.py]: Since most installations will not have GPUs, can you make this entire section conditional?
```suggestion
        if cuda_visible_devices or local_gpus:
           num_local_gpus = self.get_local_gpu_count()
           ...
```

File: lib/galaxy/jobs/runners/local.py [hunk 6]
```diff
@@ -49,23 +49,53 @@ def __init__(self, app, nworkers):
...
+
+        if num_local_gpus > 0:
+            if cuda_visible_devices:
+                cvd_list = cuda_visible_devices.split(",")
+                if num_local_gpus < len(cvd_list):
+                    cvd_list = cvd_list[:num_local_gpus] #get the first num_local_gpus gpu indices if num_local_gpus is less than the length of visible devices specified
+                    cuda_visible_devices = ",".join(cvd_list)
+            elif local_gpus:
+                if num_local_gpus < int(local_gpus):
+                    local_gpus = num_local_gpus
```
Comment (nuwang) [inline lib/galaxy/jobs/runners/local.py]:
```suggestion
                local_gpus = min(local_gpus, num_local_gpus)
```

File: lib/galaxy/jobs/runners/local.py [hunk 7]
```diff
@@ -49,23 +49,53 @@ def __init__(self, app, nworkers):
...
+                cvd_list = range(local_gpus)
```
Comment (nuwang) [inline lib/galaxy/jobs/runners/local.py]: My understanding was that the cuda visible device list needs to be a list of UUID strings, not just numbers: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-enumeration
Maybe using a technique like this to find the GPU with lowest utilization would do the trick? https://stackoverflow.com/a/69904259

File: lib/galaxy/tools/evaluation.py [hunk 8]
```diff
@@ -513,13 +513,13 @@ def _build_config_files(self):
...
             directory = ensure_configs_directory(self.local_working_directory)
-            with tempfile.NamedTemporaryFile(dir=directory, delete=False) as temp:
-                config_filename = temp.name
             if filename is not None:
-                # Explicit filename was requested, this is implemented as symbolic link
-                # to the actual config file that is placed in tool working directory
+                # Explicit filename was requested, needs to be placed in tool working directory
                 directory = os.path.join(self.local_working_directory, "working")
-                os.link(config_filename, os.path.join(directory, filename))
+                config_filename = os.path.join(directory, filename)
```
Comment (nuwang) [inline lib/galaxy/tools/evaluation.py]: This logic is no longer the same. The `os.link` call to the temp file is missing? Why was this change necessary?

**Review by nuwang — DISMISSED** [review]
Thanks for making these changes. This is on the right track I think, but needs some changes. Please use `git rebase` to rebase the code and resolve conflicts.

**Comment by mvdbeek** [comment]
I'm sorry, but I don't think this is on the right track. There's little to no point in adding support to the local runner. What we need is support on the tool framework level for specifying resource requirements, these can then be matched against destinations. You can orient yourself on https://github.com/common-workflow-language/cwltool/pull/1581/ for the required attributes that we'd have to add to the tool xsd (See #5378 for an example of how to add these).

**Comment by gulsumgudukbay** [comment]
> I'm sorry, but I don't think this is on the right track. There's little to no point in adding support to the local runner. What we need is support on the tool framework level for specifying resource requirements, these can then be matched against destinations. You can orient yourself on [common-workflow-language/cwltool#1581](https://github.com/common-workflow-language/cwltool/pull/1581) for the required attributes that we'd have to add to the tool xsd (See #5378 for an example of how to add these).

Thank you for the comment. I have been trying to learn and understand the CWL tool and its usage in Galaxy. As I understand, it is currently used only in containerized tools. I could not find anywhere in Galaxy source code that CWL is used for package tools (that use local executable). If it is used for local executable tools, could you please show me? 

Also by "orient" do you mean to use CWL (like cores_min variable), or to use an XML tag like the Galaxy SLOTS variable and then parse it in Galaxy source code? 

Previously my approach was like that, i.e. I was using a "requirement" of type "compute" and name "GPU" and specifying the GPU ID(s) via the "version" tag for the tool XML files. I was parsing it in local.py.
Should I do CWL version of that (adding a GPU resource XSD tag to the Galaxy XSD, then parsing it in the Galaxy source code, possibly in requirements.py)?

**Comment by gulsumgudukbay** [comment]
I started to add 'CUDARequirement' to the SUPPORTED_TOOL_REQUIREMENTS variable and added import cwltool.cuda to cwltool_deps.py in case I should continue by using CWL only (instead of my first approach that uses the requirement of type compute.

**Comment by mvdbeek** [comment]
> Thank you for the comment. I have been trying to learn and understand the CWL tool and its usage in Galaxy. As I understand, it is currently used only in containerized tools

I am not sure how you've reached that conclusion. I don't think there's a relationship to containers.

> I could not find anywhere in Galaxy source code that CWL is used for package tools (that use local executable)

That's not what I was asking for, for a first round we need a description of what a tool requires.
The runtime environment is not part of this and is not typically under control of Galaxy, so there's no use adding that. What we need is a way to say "this tool requires a cetain cuda version", not "We have this GPU".

> Should I do CWL version of that (adding a GPU resource XSD tag to the Galaxy XSD, then parsing it in the Galaxy source code, possibly in requirements.py)?

Yes, I think that would be a good start

> I started to add 'CUDARequirement' to the SUPPORTED_TOOL_REQUIREMENTS variable and added import cwltool.cuda to cwltool_deps.py in case I should continue by using CWL only (instead of my first approach that uses the requirement of type compute.

No, that's not the right direction. We may eventually have a job prolog that can parse these things, but I wouldn't start with that.

**Comment by mr-c** [comment]
> You can orient yourself on [common-workflow-language/cwltool#1581](https://github.com/common-workflow-language/cwltool/pull/1581) for the required attributes that we'd have to add to the tool xsd

FYI, the latest version of the CWL `CUDARequirement` proposal is at https://github.com/common-workflow-language/cwltool/blob/92c4073bf2e7965587d62abbf08914e0735e5419/cwltool/extensions-v1.1.yml#L68

The difference from https://github.com/common-workflow-language/cwltool/pull/1581 being 

1. `cudaComputeCapabilityMin` is now `cudaComputeCapability`: a single value is a minimum capability, or if a list of values is specifed then only GPUs with compute capabilities in the list should be considered.
2. `deviceCountMin` was renamed `cudaDeviceCountMin` and can also be an expression in addition to an `int`. Likewise for `deviceCountMax` becoming `cudaDeviceCountMax`

**Comment by mvdbeek** [comment]
This is looking good @gulsumgudukbay, do you want to update the xsd with the latest changes ?

**Comment by gulsumgudukbay** [comment]
> This is looking good @gulsumgudukbay, do you want to update the xsd with the latest changes ?

Yes of course, sorry for the delay, I had several paper deadlines. I will focus in it this week.

**Comment by mr-c** [comment]
another addendum from the CWL side, we are thinking about adding minimum GPU memory requirement as well: https://github.com/common-workflow-language/cwltool/issues/1744 (like the `gpu_memory_min` proposed in this PR)

**Comment by gulsumgudukbay** [comment]
> 

That is great, thanks for the comment. Will there be a maximum GPU memory limit too? Generally, GPU tools tend to use up the maximum available GPU memory. However, if we think about running multiple instances of a single tool, or multiple GPU tools at the same time, the GPU memory gets divided up. Would it be better to have more control over how the memory is divided or is it better to give the GPU that control?

**Comment by gulsumgudukbay** [comment]
> another addendum from the CWL side, we are thinking about adding minimum GPU memory requirement as well: [common-workflow-language/cwltool#1744](https://github.com/common-workflow-language/cwltool/issues/1744) (like the `gpu_memory_min` proposed in this PR)

That is great, thanks for the comment. Will there be a maximum GPU memory limit too? Generally, GPU tools tend to use up the maximum available GPU memory. However, if we think about running multiple instances of a single tool, or multiple GPU tools at the same time, the GPU memory gets divided up. Would it be better to have more control over how the memory is divided or is it better to give the GPU that control?

**Comment by mvdbeek** [comment]
This is looking good. Can you extend test/functional/tools/resource_requirements.xml with the new requirements and extend test/unit/tool_util/test_parsing.py to make sure these are available ?

**Comment by gulsumgudukbay** [comment]
> This is looking good. Can you extend test/functional/tools/resource_requirements.xml with the new requirements and extend test/unit/tool_util/test_parsing.py to make sure these are available?

I finished adding test extensions. Please let me know about the next steps, I can do them as fast as possible.

**Review by mvdbeek — APPROVED** [review]
Awesome, thanks so much @gulsumgudukbay!

**Comment by gulsumgudukbay** [comment]
> Awesome, thanks so much @gulsumgudukbay!

Thank you for all the help!
